torch/_inductor/select_algorithm.py

```diff
@@ -1125,215 +1125,237 @@ def __call__(
-            ):
-                return no_op
-
-            precompile_key = (
-                f"{name}: {inputs_key} : {torch.get_float32_matmul_precision()}"
-            )
-            if precompile_func := self.precompile_cache.get(precompile_key):
-                return precompile_func
-
-            log.info(
-                "Multithreaded precompilation for %d choices using %d worker threads",
-                len(choices),
-                num_workers,
+                None,
             )
 
-            # In rare circumstances, because python threads inherit global state,
-            # thread pool executor can race and leave stdout/stderr in a state
-            # different than the original values. we explicitly restore the state
-            # here to avoid this issue.
+            if cached_timings == {}:
+                if not (config.max_autotune_gemm or config.max_autotune):
+                    selected_choice = choices[0]
 
-            initial_stdout = sys.stdout
-            initial_stderr = sys.stderr
+                    log.debug("One or more missing cached autotunings, selecting default %s.", selected_choice)
 
-            def precompile_with_captured_stdout(choice):
-                with restore_stdout_stderr(initial_stdout, initial_stderr):
-                    return choice.precompile()
+                    wait_on_precompile = precompile([selected_choice])
+                    wait_on_precompile()
 
-            executor = ThreadPoolExecutor(max_workers=num_workers)
+                    return selected_choice.output_node
+                else:
+                    log.debug("One or more missing cached autotunings, proceeding to autotuning.")
+            else:
+                selected_choice = min(cached_timings, key=cached_timings.__getitem__)
 
-            futures = {}
-            for c in choices:
-                if hasattr(c, "precompile"):
-                    future = executor.submit(precompile_with_captured_stdout, c)
-                    futures[future] = c
+                log.debug("Selecting choice %s from cached autotunings.", selected_choice)
 
-            @functools.lru_cache(None)
-            @restore_stdout_stderr(initial_stdout, initial_stderr)
-            def wait_on_futures():
-                counters["inductor"]["select_algorithm_precompile"] += 1
-                for future in as_completed(
-                    futures,
-                    timeout=precompilation_timeout_seconds,
-                ):
-                    if e := future.exception():
-                        log.error(
-                            "Exception %s for benchmark choice %s", e, futures[future]
-                        )
+                wait_on_precompile = precompile([selected_choice])
+                wait_on_precompile()
 
-                executor.shutdown(wait=True)
+                return selected_choice.output_node()
+        
+        assert config.max_autotune_gemm or config.max_autotune
```

Comment:
another good check to have imo, we shouldn't cross this boundary unless we're autotuning